rfcs/337-daily-ci-cache-rebuilds.md

```diff
@@ -0,0 +1,90 @@
+avoid having to re-download languages on every run) as well as language specific
+cache (e.g. golangci-lint cache). Importing these requires `checksum`, a CircleCI
+function to calculate the checksum of a file, to be used to ensure the correct
+version of each language is being used, which leads to a lot of extra steps in the
+cache generation process.
+
+### Example: Golang Version Calculation
+
+Currently there are 3 steps to calculating the golang version, as we can see below:
+
+![go version calculation](./337-daily-ci-cache-rebuilds/go-cache-key-calc.png)
+
+This is a lot of extra steps that ends up making the cache generation process
+take up a lot of space in the CI pipeline, making it harder for users to understand
+what's going on and separate it from the business logic of the pipeline.
+
+### Goals
+
+- Reduce the time it takes to generate caches
+- Reduce the time it takes to consume caches
+- Reduce the amount of dependencies fetched outside of the cache (i.e. asdf
+  languages)
+- Reduce the size of the caches, as much as possible (without sacrificing
+  the above goals)
+
+## Design Details
+
+The proposed solution is to generate the caches daily and consume them in each CI job.
+
+This will be done by creating a new `generate-cache` workflow that runs a job creating
+the cache required for each job (asdf, go cache, etc.) then uploading them to CircleCi
+using the `save_cache` command in the format of `v1-daily-cache-{{ timestamp }}`.
+
+These caches will be imported by using the `restore_cache` command, which will restore
+the newest cache that matches the key `v1-daily-cache` as per
+[CircleCI `restore_cache` docs](https://circleci.com/docs/caching#restoring-cache).
+
+## Testing and Release Plan
+
+Run on CI for a week on two internal Outreach services to ensure that the
+caching strategy is working as expected. We could potentially create a job
+that runs a cache workflow, using [setup workflows](https://circleci.com/docs/dynamic-config)
+to run the cache workflow on a PR and ensures that it caches the paths
+expected.
+
+## Implementation History
+
+- 2022-10-03: Initial draft
+
+## Drawbacks
+
+Generating caches daily creates the possibility of a cache being out of date
+for a period of time. This is mitigated by the fact that `devbase`, and other tools,
+will fetch things not inside of the cache. This causes a temporary slowdown, but is
+only a problem for a day. This is further mitigated by being able to re-trigger the
+cache generation process anytime.
+
+## Alternatives
+
+- Keep the current cache generation process, and just fix the cache generation to
+  properly include all directories required for caching (e.g. go cache). This would
+  help make the cache more efficient, but it would not solve the problem of the asdf
+  version cache being bloated over time without introduction of more dynamic cache
+  key computation and would continue to bloat CI logs.
```

Comment:
I really like keeping the current process. I think generating the cache daily could be problematic, as you outlined.

What about keeping both options. Upload a new cache if things change in a job but also use a daily cache run for pruning. The daily flow would stay the same as you described, but the goal would change to removing things that we won't need for future runs. We could even reduce the frequency of this since it's only goal is to reduce bloat.

Reply:
@artemchernyak-outreach I don't think I like that. Mainly because cache generation bloat is still in the CI steps display, and now we have the added complexity of having to run the steps to regenerate the cache in two different places (pruning _is_ regenerating it, ultimately). I think being able to trigger it whenever you want solves pretty much all of the issues around invalidating it, which is what doing it on each job kinda does for you.

Reply:
Talked about this during standup. And came to the conclusion:

If we only run the daily job conditionally (based on new changes to the repo) we will proceed with the implementation of the daily job to update the cache.